File: apps/betty-server/src/trips/application/use-cases/end-trip/end-trip.use-case.ts
import { Injectable, Inject } from '@nestjs/common';
import { TripRepository } from '../../../domain/repositories/trip.repository';
import { Trip } from '../../../domain/entities/trip.entity';
import { TRIP_REPOSITORY } from '../../../infrastructure/ioc/symbols';

@Injectable()
export class EndTripUseCase {
  constructor(
    @Inject(TRIP_REPOSITORY)
    private readonly tripRepository: TripRepository
  ) {}

  async execute(id: string): Promise<Trip> {
    const trip = await this.tripRepository.findById(id);

    if (!trip) {
      throw new Error('Trip not found');
    }

    if (trip.isCompleted) {
      throw new Error('Trip is already completed');
    }

    return this.tripRepository.end(id);
  }
}

File: apps/betty_app/lib/trips/application/use_cases/end_trip_use_case.dart
import '../../infrastructure/services/trip_api_service.dart';
import '../../domain/entities/trip.dart';

class EndTripUseCase {
  final TripApiService _apiService;

  const EndTripUseCase(this._apiService);

  Future<Trip> execute(String id) async {
    return await _apiService.endTrip(id);
  }
}
